### PR TITLE
implement delete

### PR DESCRIPTION
**Pull request by whyrusleeping**

### CODE REVIEW

File: amt.go [hunk 1]
```diff
@@ -309,6 +363,20 @@ func (n *Node) setBit(i uint64) {
...
+func (n *Node) clearBit(i uint64) {
+	if i > 7 {
+		panic("cant deal with wider arrays yet")
+	}
+
+	if len(n.Bmap) == 0 {
+		panic("invariant violated: called clear bit on empty node")
+	}
+
+	mask := byte(((1 << 8) - 1) - (1 << i))
```
Comment (magik6k) [inline amt.go]: `(1 << 8) - 1` is a weird way of typing `0xff`

File: amt.go [hunk 2]
```diff
@@ -200,14 +203,65 @@ func (n *Node) get(bs Blocks, height int, i uint64, out interface{}) error {
...
 	return subn.get(bs, height-1, i%nodesForHeight(width, height), out)
 }
 
+func (r *Root) Delete(i uint64) error {
+	if err := r.Node.delete(r.bs, int(r.Height), i); err != nil {
+		return err
+	}
+	r.Count--
+
+	for r.Node.Bmap[0] == 1 {
```
Comment (magik6k) [inline amt.go]: Something about this `1` feels at least counter-intuitive. Is it that if only 1 bit is set then we unpack the level below? Shouldn't that be doing `bits.OnesCount == 1` then? (and `bits.TrailingZeros` to get the index later)

Comment (magik6k) [reply amt.go]: (https://golang.org/pkg/math/bits/, also, both of those calls should get compiled down to single x86 instructions (`popcnt`, `tzcnt`))

File: amt.go (hunk 1 above)
Comment (whyrusleeping) [reply amt.go]: >.>

File: amt.go (hunk 2 above)
Comment (whyrusleeping) [reply amt.go]: This is specifically if the lowest bit is set only.

Looking at this though, I think that it doesn't account for the lowest level having only a single item in that slot.

**Pull request merged**